pupil_src/shared_modules/task_lib/interface.py

```diff
@@ -0,0 +1,190 @@
+        raise exceptions.
+        """
+        return self._completed
+
+    @property
+    def ended(self):
+        """
+        True if the task ended in any way. Either successfully, or it was canceled,
+        or it raised an exception.
+        """
+        return self.completed or self.canceled_or_killed
+
+    @property
+    def canceled_or_killed(self):
+        """
+        True if the task was terminated from the foreground.
+        """
+        return self._canceled_or_killed
+
+    @abc.abstractmethod
+    def start(self):
+        """
+        Start the task.
+
+        Make sure to add any observers before calling this.
+
+        Raises:
+            ValueError: If the task was already started.
+        """
+        if self.started:
+            raise ValueError("Task already started!")
+        self._started = True
+
+    @abc.abstractmethod
+    def cancel_gracefully(self):
+        """
+        Ask the background process to shut itself down at the next occasion.
+
+        For generator functions, this will be after the next yield.
+        Routines cannot be interrupted, so they will just terminate normally.
+
+        After this was called, any results by the process are ignored. You will not
+        get notified about any yields, return values, or exceptions.
+        However, on_canceled_or_killed() will be triggered when the process terminates.
+
+        Compared to kill():
+        - The task is not terminated immediately and there are no guarantees that it
+          gets terminated at all. However, all processes are started as daemons,
+          so at least they terminate when the main process terminates.
+        - The foreground process does not block.
+        - The background can clean up normally and will not break shared resources etc.
+        - on_canceled_or_killed() will be called in the next update() after the task
+          terminated.
+
+        Raises:
+            ValueError: If the task is not running, i.e. it was never started or it
+                already ended.
+        """
+        if not self.running:
+            raise ValueError("Task not running, can only cancel running tasks!")
+
+    @abc.abstractmethod
+    def kill(self, grace_period):
+        """
+        Try to cancel the task gracefully and kill it if that's not successful.
+
+        Compared to cancel_gracefully():
+        - You can be sure that the task gets terminated during the grace
+          period or immediately after.
+        - The foreground process blocks until the process terminates or the grace
+          period is over.
+        - You likely break resources shared between your background process and the
+          foreground.
+        - You can be sure that on_canceled_or_killed() gets called immediately and not
+          during the next update().
+
+        Args:
+            grace_period (Number): Seconds to wait for the process to shut itself
+                down. If None, there is no grace period.
```

Comment:
Technically yes, with `0` it will trigger a `cancel_gracefully`, with `None` it doesn't. Practically, it doesn't make a difference. I thought that `kill(grace_period=None)` reads better.